---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  echo = TRUE,
  message = FALSE,
  error = FALSE,
  comment = "#>",
  fig.path = "man/figures/"
)
```

# FacileArchs4Data

Implements facile access over local copies of [ARCHS4][] (v2) datasets.
Quickly slice and dice samples and studies, and effortlessly retrieve expression
data to perform analyses with the [FacileAnalysis][facileanalysis] or standard
Bioconductor workflows.

**This package is a work in progress.**

[arcsh4]: https://archs4.org/
[facileanalysis]: https://github.com/facilebio/FacileAnalysis

You will need to install the package, download the archs4 hdf5 files, and
configure this package for use.

## Installation


Install the package from GitHub

```r
pak::pak("facilebio/FacileArchs4Data")
library(FacileArchs4Data)
```

## Download Data

You'll need to [download](archs4dl) the archs4 hdf5 files to a folder on your
local filesystem (`ARCHS4DIR`). Rename these files to include the date of
their release (highglighted yellow here from their [download](archs4dl) page):

[![archs4dlimage](man/figures/archs4-dl-page.png)

Let's say:

```bash
export ARCHS4DIR="~/workspace/data/archs4"
cd $ARCHS4DIR
curl -O https://s3.dev.maayanlab.cloud/archs4/files/mouse_gene_v2.latest.h5
curl -O https://s3.dev.maayanlab.cloud/archs4/files/mouse_gene_v2.latest.h5
```

more complex

```bash
export ARCHS4DIR="~/workspace/data/archs4"
cd $ARCHS4DIR
curl - \
  2025-11-11__mouse_gene_v2.latest.h5 \
  https://s3.dev.maayanlab.cloud/archs4/files/mouse_gene_v2.latest.h5
curl -o \
  2025-11-11__human_gene_v2.latest.h5 \
  https://s3.dev.maayanlab.cloud/archs4/files/mouse_gene_v2.latest.h5
```

Note that the date of release listed on the file doesn't not always match the
date that is "baked into" the hdf5 file itself. After downloading these files
you can call `FacileArchs4Data::archs4_dir_info("~/workspace/data/archs4") and
look at the `creation_date` column. The files that were released on 2025-11-11
have an internal creation date of `2025-11-06`.

[archs4dl]: https://archs4.org/download

## Configuraton

Your life will be easier if you set the `archs4.data_dir` R option to your
`$ARCHS4DIR`. To do so, you can add this line to your `~/.Rprofile` file:

```r
options(archs4.data_dir = "~/workspace/data/archs4")
```

An easy way to open your `~/.Rprofile` file is to use the {usethis} package:

```r
usethis::edit_r_profile()
```

For snappier slicing and dicing into these hdf5 files, there are some cache
files to creat.

## Create Cache Files

Along with the gene-level counts assay matrix, the hdf5 files also store
the sample level metadata for the ncbi/geo samples along with gene-level
metadata.

To make slicing and dicing data out of the hdf5 quicker, we pull out this
metadata to parquet files.

Running the command below will, by default, create a `${ARCHS4DIR}/cache`
directory, and serialize parquet files into there for each of the arcsh4
hdf5 (`*.h5`) files  in `${ARCHS4DIR}`

```r

```

```r
# This wil default to getOption("archs4.data_dir"), but just being
# explicit here
archs4_create_sample_cache("~/workspace/data/archs4")
```

## Usage

Currently you should use the `a4 <- Archs4Client$new("mouse")` (or `"human"`) R6
constructor to get a thin wrapper around the data object. Then, subset the
`a4$samples` tibble to include the the samples you want to analyze. Finally,
convert the sample-frame it to a `DGEList` or `SummarizedExperiment` to move on
with your life -- see the **Usage** section for details.

To make your life easy, store the "equally versioned" mouse and human
datasets in a single directory (`$ARCHS4_DATADIR`) with the value of `species`
in the filename, and set a global R option (`archs4.data_dir`) to
`$ARCHS4_DATADIR`.

You can put new/old h5 files in different directories, then load the dataset by
specifying simply the species. See the **Installation** section for more
details.

Note that when you load the hdf5 file, a subset of columns will be loaded
for all of the samples by default. This is the (likely few second delay) you
experience when calling the `$new()` constructor for the first time.


```{r, message=FALSE, warning=FALSE}
library(FacileArchs4Data)
library(tidyverse)
theme_set(theme_bw(base_size = 16))
# NOTE: You can set this in your ~/.Rprofile file to make your life easier
options(archs4.data_dir = "~/workspace/data/archs4/v2.6")
a4h <- Archs4Client$new("human")
```

## Finding samples / expreiments of interest

Some of the funtions in this package are fashioned after the archs4py package,
so there is a `$search()` function made available to you that enables you to
search the metadata of each sample based on some of the decault columns.

Let's say you were interested in searching for "parksinsons" related samples:

```{r}
samples.park <- a4h$search("parkinson")
```

This will return a tibble of samples where the word `"parkinson"` appears
in one of hte meta data fields:

```{r}
head(samples.park)
```

The sample metadata tibble is also easily available in the `$samples` tibble,
and you can search for rows of interest "as you are used to".

## Expression Analysis

Let's say you already knew you were interested in using the `"GSE40710"` dataset
(aSyn polyA-RNAseq in PD and unaffected cortical brain samples) to explore the
gene expression differencs between healthy and parkinson's brain transcriptomes.

We can easily extract the samples in this dataset like so:

```{r}
samples <- filter(a4h$samples, series_id == "GSE40710")
```

Your next challenge will be to identify which samples belong to healthy vs
affected individuals. The data in archs4 matches what is provided in the GEO
database, and there are unfortunately no standards as far as data/sample
annotations go, so each individual experiment will provide you with its own
set of challenges.

Often, the sample-level metadata you will be most interested will be in the
`"source_name_ch1"` and `"title"` columns:

```{r}
set.seed(0xBEEF)
samples |>
  sample_n(6) |>
  select(sample, title, source_name_ch1)
```

We can see that the parkinson's vs control status is in the `"title"` column.
Let's great a "group" column with the affected status, and then use the
`biocbox` function in this package to retrieve the associated expression data
for these samples and turn all of it into a DGEList for downstream analysis.

```{r}
y <- samples |>
  mutate(
    # extract control/parksinson & remove the title column
    group = sub(" .*", "", title),
    # the "'" to make `group` a well behaved R variable
    group = sub("'", "", group)
  ) |>
  biocbox(class = "DGEList")
```

Now you are ready to analyze `y` using your favorite edgeR/limma workflow.

## FacileData Integration

If you have the FacileBiocData and FacileAnalysisShine package installed, you
can then use `y` for interactive analysis.

```{r, message=FALSE, warning=FALSE}
library(FacileBiocData)
library(FacileAnalysisShine)
yf <- facilitate(y)
```

### Principal Components Analysis

```{r}
ypca <- fpca(yf)
viz(ypca, color_aes = "group")
```

### Differential Expressoin Analysis

```{r}
ydge <- yf |>
  flm_def(covariate = "group", numer = "Parkinsons", denom = "Control") |>
  fdge(method = "voom")
```

```{r}
ydge |>
  tidy() |>
  ggplot(aes(x = logFC, y = -log10(padj))) +
  geom_point(alpha = 0.8) +
  xlim(-3, 3)
```

Or if you're webgl-powered:

```{r}
viz(ydge, type = "volcano", ntop = 2000)
```


Refer to the [FacileAnalysis][] documentation for more analysis fun.

## License

Please note that the use of archs4 data is subject to its own agreement /
licensing restrictions. Be sure to follow those.

The code in this package is released under the [Apache 2 License][apache2].

## Acknowledgements

Thanks to Alexander Lachmann and The Maâ€™ayan Laboratory for the development
and continued updates to the [ARCHS4 project][archs4].

[FacileAnalysis]: https://facilebio.github.io/FacileAnalysis/
[facilebio]: https://facile.bio/
[apache2]: https://www.apache.org/licenses/LICENSE-2.0
